src/Http/Middleware/CsrfProtectionMiddleware.php

```diff
@@ -223,16 +240,57 @@ protected function _validateToken(ServerRequestInterface $request): void
         $body = $request->getParsedBody();
         if (is_array($body) || $body instanceof ArrayAccess) {
             $post = Hash::get($body, $this->_config['field']);
-            if (Security::constantEquals($post, $cookie)) {
+            if ($this->_compareToken($post, $cookie)) {
                 return;
             }
         }
 
         $header = $request->getHeaderLine('X-CSRF-Token');
-        if (Security::constantEquals($header, $cookie)) {
+        if ($this->_compareToken($header, $cookie)) {
             return;
         }
 
         throw new InvalidCsrfTokenException(__d('cake', 'Missing CSRF token body'));
     }
+
+    /**
+     * Ensure that the request token matches the cookie value and that
+     * both were generated by us.
+     *
+     * @param mixed $post The request token.
+     * @param mixed $cookie The cookie token.
+     * @return bool
+     */
+    protected function _compareToken($post, $cookie): bool
+    {
+        if (!is_string($post)) {
+            $post = '';
+        }
+        if (!is_string($cookie)) {
+            $cookie = '';
+        }
+        $postKey = (string)substr($post, 0, static::TOKEN_VALUE_LENGTH);
+        $postHmac = (string)substr($post, static::TOKEN_VALUE_LENGTH);
+        $cookieKey = (string)substr($cookie, 0, static::TOKEN_VALUE_LENGTH);
+        $cookieHmac = (string)substr($cookie, static::TOKEN_VALUE_LENGTH);
+
+        // Put all checks in a list
+        // so they all burn time reducing timing attack window.
+        $checks = [
+            hash_equals($postKey, $cookieKey),
```

Comment:
Wouldn't this always be the same if both $post and $cookie are ''?

Reply:
Yes it would. But we need to run all the same comparisons to reduce timing attack windows.

Reply:
I mean won't it be a false positive?

Reply:
`_validateToken()` has `if (!$cookie)` check for empty cookie.

But check for cookie/post minimal length could be done too.

Reply:
If they aren't long enough, substr can generate false which is cast to string which could be the same also.

Reply:
You would somehow need a key whose hmac matches `(string)false` which doesn't seem feasible.  Looks good to me.